src/fides/api/service/connectors/dynamic_erasure_email_connector.py

```diff
@@ -0,0 +1,258 @@
+import json
+
+from typing import Any, Dict, List, Optional
+
+from loguru import logger
+from sqlalchemy import and_
+from sqlalchemy.orm import Query, Session
+
+from fides.api.common_exceptions import MessageDispatchException
+from fides.api.graph.config import GraphDataset, CollectionAddress
+from fides.api.graph.execution import ExecutionNode
+from fides.api.models.connectionconfig import ConnectionConfig
+from fides.api.models.datasetconfig import DatasetConfig
+from fides.api.models.privacy_request import (
+    ExecutionLog,
+    ExecutionLogStatus,
+    RequestTask,
+    PrivacyRequest,
+    TraversalDetails,
+)
+from fides.api.models.policy import Policy
+from fides.api.schemas.connection_configuration.connection_secrets_dynamic_erasure_email import (
+    DynamicErasureEmailSchema,
+)
+from fides.api.schemas.policy import ActionType
+from fides.api.service.connectors.base_connector import BaseConnector
+from fides.api.service.connectors.erasure_email_connector import (
+    GenericErasureEmailConnector,
+    filter_user_identities_for_connector,
+    send_single_erasure_email,
+)
+from fides.config import get_config
+
+CONFIG = get_config()
+
+
+class DynamicErasureEmailConnector(GenericErasureEmailConnector):
+    """Email Erasure Connector that performs a lookup for the email to use."""
+
+    config_schema = DynamicErasureEmailSchema
+
+    def __init__(self, configuration: ConnectionConfig):
+        super().__init__(configuration)
+
+    def get_email_address_from_custom_request_fields(
+        self,
+        connector: BaseConnector,
+        graph_dataset: GraphDataset,
+        privacy_request: PrivacyRequest,
+        collection_address: CollectionAddress,
+        field_name: str,
+    ) -> Optional[str]:
+        """
+        Gets the custom request fields from the privacy request and uses them to perform a
+        standalone query on the connector, in order to retrieve the value of the email address
+        we need to send the erasure email to (specified in field_name).
+
+        Returns the found email address, or None if none were found. Expects that the query using
+        the provided custom request fields returns 1 single row, otherwise an error is logged and
+        None is returned.
+        """
+        # The following code is a bit hacky, but we need to instantiate a RequestTask
+        # and an Execution node in order to execute a standalone retrieval query on the connector.
+        # The only fields we really care about in the RequestTask are those related to the collection
+        # and to the dataset itself.
+
```

Comment:
note to self: I think I probably need to add some exception handling in this function